New function 'sampleRecords'

wgaul opened the pull request:
Hi, 

I’m a PhD student under Dr. Jon Yearsley at University College Dublin, Ireland.  I’m using ‘virtualspecies’ for a simulation study on methods for correcting spatial sampling bias in biological records datasets.  I’ve written a function called ‘sampleRecords’, largely based on the existing ‘sampleOccurrences’ function, but with some modifications that produce sample data similar to the biological records datasets I am using.  I think this function would be useful to others.  Please take a look and let me know what you think – I am happy to discuss and make changes if you have ideas.  

The purpose of this function is to sample virtual species occurrences with replacement to allow the possibility of multiple observations from each site.  This function produces a data frame of observations in which each row is a species record, with the coordinates of the observation (see examples in the documentation).  This produces data that is similar to many opportunistic biological records datasets, (e.g. datasets available on GBIF) which may contain many records of the same species from some locations and no or few records from other locations.  This function will be useful for simulation studies exploring the effect of sampling bias, and will also be useful for simulation studies using hierarchical occupancy/detection models (which require replicate site visits to estimate detection probability).  I think the ‘sampleOccurrences’ function was not able to produce multiple replicate samples from a cell, so this ‘sampleRecords’ function extends the capabilities of the ‘virtualspecies’ package by producing sample data with replicate records.  

Also, a minor note is that I used ‘roxygen2’ to generate the documentation, and it seems that it automatically made some small changes to most of the ‘man’ files, including removing some empty lines and shuffling the order of some blocks.  I don’t think these changes will hurt anything, but I couldn’t figure out how to stop them from happening.  Every time I re-generated the documentation those changes happened.  Let me know if this causes problems, and I can try to get rid of all those small changes and re-submit a pull request.

All the Best,
Willson

wgaul committed:
Created new function to sample n records rather than n cells.  This allows simulating replicate samples from cells.

wgaul committed:
Created new function to sample n records rather than n cells.  This allows simulating replicate samples from cells.

wgaul committed:
Created new function to sample n records rather than n cells.  This allows simulating replicate samples from cells.

wgaul committed:
Created new function to sample n records rather than n cells.  This allows simulating replicate samples from cells.

wgaul committed:
Added comments about how the function works to help me understand it before I start modifying code.

wgaul committed:
Added comments about how the function works to help me understand it before I start modifying code.

wgaul committed:
In sampleRecords() function, changed to use rbinom to generate observations in the presence-absence case. Added comments to help me follow the code flow.

wgaul committed:
In sampleRecords() function, changed to use rbinom to generate observations in the presence-absence case. Added comments to help me follow the code flow.

wgaul committed:
Un-commented the old way of using dismo::randomPoints to sample points so that the code runs (I think it won't run in the previous commit).

wgaul committed:
Un-commented the old way of using dismo::randomPoints to sample points so that the code runs (I think it won't run in the previous commit).

wgaul committed:
For the presence-absence case, sampleRecords() now samples locations with replacement and returns a data frame of observations at locations.

wgaul committed:
For the presence-absence case, sampleRecords() now samples locations with replacement and returns a data frame of observations at locations.

wgaul committed:
Plot of sampled observations and true p/a distribution now saved as an element of the result list returned by sampleRecords().  Not sure if I should keep this, or might need to change the str method to keep console output manageable.

wgaul committed:
Plot of sampled observations and true p/a distribution now saved as an element of the result list returned by sampleRecords().  Not sure if I should keep this, or might need to change the str method to keep console output manageable.

wgaul committed:
Presence-only case now working in sampleRecords().  Also moved the line saving the plot in results list so that it saves for both presence-only and p/a plots.

wgaul committed:
Presence-only case now working in sampleRecords().  Also moved the line saving the plot in results list so that it saves for both presence-only and p/a plots.

wgaul committed:
Added .Random.seed to attributes of results.  Added documentation and an example showing how to reproduce results based on the stored .Random.seed.

wgaul committed:
Added .Random.seed to attributes of results.  Added documentation and an example showing how to reproduce results based on the stored .Random.seed.

wgaul committed:
Changed to resample with replacement and return records for the simplest default case (no bias, presence-only). Updated example for that case.

wgaul committed:
Changed to resample with replacement and return records for the simplest default case (no bias, presence-only). Updated example for that case.

wgaul committed:
Changed presenc-absence no bias case to resample with replacement.

wgaul committed:
Changed presenc-absence no bias case to resample with replacement.

wgaul committed:
Added example of how to view saved results plot.  Modified the correct.by.suitability case to work with sample.points as a SpatialPointsDataFrame. Switched to rbinom for random sampling.

wgaul committed:
Added example of how to view saved results plot.  Modified the correct.by.suitability case to work with sample.points as a SpatialPointsDataFrame. Switched to rbinom for random sampling.

wgaul committed:
Updated examples in documentation.

wgaul committed:
Updated examples in documentation.

wgaul committed:
Starting to work on sample prevalence.  Discovered a small bug in the sampleOccurrences() code, so am going to try to fix that on a different branch and then merge the changes into this branch.

wgaul committed:
Starting to work on sample prevalence.  Discovered a small bug in the sampleOccurrences() code, so am going to try to fix that on a different branch and then merge the changes into this branch.

wgaul committed:
Fixed bug when specifying sample.prevalence in sampleRecords() with bias.

wgaul committed:
Fixed bug when specifying sample.prevalence in sampleRecords() with bias.

wgaul committed:
Updated documentation about potential problems when specifying sample.prevalence and bias at the same time.

wgaul committed:
Updated documentation about potential problems when specifying sample.prevalence and bias at the same time.

wgaul committed:
Merge branch 'master' into sample_records

wgaul committed:
Made sample.points into SpatialPointsDataFrame in the case of sample.prevalence with bias.

wgaul committed:
Made sample.points into SpatialPointsDataFrame in the case of sample.prevalence with bias.

wgaul committed:
Finished switching all cases to resample with replacement.  Cleand up comments.

wgaul committed:
Finished switching all cases to resample with replacement.  Cleand up comments.

wgaul committed:
Fixed bug when using presence-only sampling on a map with only a single cell containing a species presence.

wgaul committed:
Fixed bug when using presence-only sampling on a map with only a single cell containing a species presence.

wgaul committed:
Tidying up comments and documentation.

wgaul committed:
Tidying up comments and documentation.

wgaul committed:
Tested the case of a species with no presences.  Sampling occurrs correctly.  Removed TODO note.

wgaul committed:
Tested the case of a species with no presences.  Sampling occurrs correctly.  Removed TODO note.

wgaul committed:
Cleaned up documentation and examples.

wgaul committed:
Cleaned up documentation and examples.

wgaul committed:
Merge pull request #1 from wgaul/sample_records

Sample records

Farewe commented:
Thanks Willson, this looks great!

 I will look into that as soon as possible.

Farewe committed:
virtualspecies version 1.4-3:
- modified function sampleOccurrences (thanks Willson Gaul).
   --> added the possibility to sample multiple records in the same cells
   --> stored the random seed in output object so that samplings can be reproduced
   --> added the possibility to sample False Positives in presence only mode
   --> improved the output list from sampleOccurrences

- added print and str generic functions for class VSSampledPoints

- minor changes to other functions

Farewe committed:
Fixed conflict in imports with dismo

Farewe committed:
Clarified how to use arguments correct.by.suitability and weights.

Farewe committed:
Fixed a bug introduced in sampleOccurrences

wgaul committed:
Resolved conflicts by keeping Farewe's .randomPoints code.

wgaul committed:
Removed sampleRecords function as this is now implemented in sampleOccurrences.

wgaul committed:
Merge branch 'master' of https://github.com/wgaul/virtualspecies

Trying to get my branch to match Farewe's master.

wgaul committed:
deleted sampleRecords documentation.  Trying to get my fork to match Farewe's.

wgaul committed:
Removed export of old unused function.

wgaul committed:
Deleted old testing file.

wgaul committed:
Removed code that set number of points to resample equal to number of cells in mask.  I think this is a bug preventing resampling with replacement when using a mask.

wgaul committed:
Stopped tracking some note-taking files.
